subt_ign/src/level_generator.cc

```diff
@@ -0,0 +1,313 @@
+#include "ConnectionHelper.hh"
+#include "SdfParser.hh"
+
+using namespace ignition;
+using namespace subt;
+
+/// \brief Data structure containing properties of a level
+struct Level
+{
+  /// \brief Position X
+  double x;
+
+  /// \brief Position Y
+  double y;
+
+  /// \brief Position Z
+  double z;
+
+  /// \brief Size x
+  double sx;
+
+  /// \brief Size y
+  double sy;
+
+  /// \brief Size z
+  double sz;
+
+  /// \brief Buffer of a level
+  double buffer;
+
+  /// \brief A list of models referenced by this level
+  std::vector<std::string> refs;
+};
+
+/// \brief A list of models to ignore when generating levels
+const std::set<std::string> modelsToIgnore =
+{
+  "staging_area", "base_station", "artifact_origin"
+};
+
+/// \brief Set to true to print debug visuals
+bool debugLevels = false;
+
+/// \brief Print debug visuals the levels sdf string
+/// \param[in] _levels List of levels to debug
+void printDebugVisuals(const std::map<double, Level> &_levels)
+{
+  std::stringstream out;
+  static int debugCounter = 0;
+
+  for (const auto &lIt : _levels)
+  {
+    auto l = lIt.second;
+    out << "    <model name=\"level_debug_" << std::to_string(debugCounter++)
+        << "\">\n";
+    out << "      <pose>" << l.x << " " << l.y << " " << l.z
+                          << " 0 0 0</pose>\n";
+    out << "      <link name=\"level_debug_link\">\n";
+    out << "        <visual name=\"level_debug_visual\">\n";
+    out << "          <transparency>0.2</transparency>\n";
+    out << "          <geometry><box><size>\n";
+    out << "            " << l.sx << " " << l.sy << " " << l.sz << "\n";
+    out << "          </size></box></geometry>\n";
+    out << "          <material><diffuse>1 0 0</diffuse></material>\n";
+    out << "          <cast_shadows>false</cast_shadows>\n";
+    out << "        </visual>\n";
+    out << "      </link>\n";
+    out << "      <static>true</static>\n";
+    out << "    </model>\n";
+  }
+  std::cout << out.str() << std::endl;
+}
+
+/// \brief Get the levels sdf string
+/// \param[in] _level input map of levels
+std::string levelsStr(const std::map<double, Level> &_levels)
+{
+  static int levelCounter = 0;
+  std::stringstream out;
+  for (const auto &lIt : _levels)
+  {
+    auto l = lIt.second;
+    out << "      <level name=\"level" << std::to_string(levelCounter++)
+        << "\">\n";
+
+    for (const auto &r : l.refs)
+    {
+      out << "        <ref>" << r << "</ref>\n";
+    }
+    out << "        <pose>" << l.x << " " << l.y << " " << l.z
+                            << " 0 0 0</pose>\n";
+    out << "        <geometry><box><size>" << l.sx << " " << l.sy << " " << l.sz
+                            << "</size></box></geometry>\n";
+    out << "        <buffer>" << l.buffer << "</buffer>\n";
+
+    out << "      </level>\n";
+  }
+  return out.str();
+}
+
+
+/// \brief Print the level sdf elements
+/// \param[in] _vertexData vector of model info
+///\ param[in[ _size Tile size
```

Comment:
Is `_size` the size of each level?

Reply:
updated doc. It should be at minimum the size of a tile and used to determine the final size of level. 500847a9e4f22861f0c737b3386ae60364523fb8